specification/logs/semantic_conventions/Instrumentation/mobile-events.md

```diff
@@ -0,0 +1,57 @@
+# Semantic Conventions for mobile instrumentation
+
+**Status**: [Experimental][DocumentStatus]
+
+This document defines semantic conventions for instrumentation on mobile platforms.
+
+<!-- toc -->
+
+- [Lifecycle instrumentation](#lifecycle-instrumentation)
+  * [`state` values](#state-values)
+    + [iOS](#ios)
+    + [Android](#android)
+
+<!-- tocstop -->
+
+## Lifecycle instrumentation
+
+This section defines how to apply semantic conventions when instrumenting application lifecycle.
+This event is meant to be used in conjunction with `os.name` [resource semantic convention](https://github.com/open-telemetry/semantic-conventions/blob/main/specification/resource/semantic_conventions/os.md) to identify platform.
+
+**Event name**: `client.lifecycle`
+
+**Event domain**: `device`
+
+The following attributes are stored in the `event.data` map.
+
+| Attribute | Type   | Description                                | Examples          | Requirement level |
+|-----------|--------|--------------------------------------------|-------------------|-------------------|
+| `state`   | String | The state entered at the time of the event | platform specific | Required          |
+
+### `state` values
+
+#### iOS
+
+The iOS lifecycle states are defined in the [UIApplicationDelegate documentation](https://developer.apple.com/documentation/uikit/uiapplicationdelegate#1656902),
+and from which the `OS terminology` column values are derived.
+
+| Name         | OS terminology                 | description                               |
+|--------------|--------------------------------|-------------------------------------------|
+| `active`     | applicationDidBecomeActive     | The app has become "active"               |
+| `inactive`   | applicationWillResignActive    | The app is about to become "inactive".    |
+| `background` | applicationDidEnterBackground  | The app is now in the background.         |
+| `foreground` | applicationWillEnterForeground | The app is about to enter the foreground. |
+| `terminate`  | applicationWillTerminate       | The app is about to terminate.            |
+
+#### Android
+
+The Android lifecycle states are defined in [Activty lifecycle callbacks](https://developer.android.com/guide/components/activities/activity-lifecycle#lc), and from which the `OS idenfifiers` are derived.
```

Comment:
We should probably use this link as a reference instead of the Activity one: https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner - While the names are similar to the ones listed for Activities, using an Activity-related reference might be misleading for the purpose of this app-wide lifecycle events definition.

Reply:
Just a point of clarification: the callbacks/events defined here aren't for the app process or even the global Android `Application`, but for [lifecycle owners](https://developer.android.com/reference/androidx/lifecycle/LifecycleOwner) like an Activity or Fragment that has an [Android lifecycle](https://developer.android.com/reference/androidx/lifecycle/Lifecycle) and dispatches [lifecycle events](https://developer.android.com/reference/androidx/lifecycle/Lifecycle.Event) and are in lifecycle [states](https://developer.android.com/reference/androidx/lifecycle/Lifecycle.State). 

Specifically, [ProcessLifecycleOwner](https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner) is an aggregation of all lifecycle owners in the application, allowing you to hook into all lifecycle owners' `onPause()`, for instance, without attaching the callback it to each lifecycle owner.

A broader question I have is about what this is trying to achieve. Do we want to add in the notion of state of the app (e.g. foreground vs backgrounded), or the state of the lifecycle owner (e.g. `Activity`) at the time of the event?

Reply:
Good point @bidetofevil - The callbacks are a composite of lifecycle owners (probably only activities though, based on how [the docs](https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner) explain how it works), however, is not exactly as attaching a callback to each lifecycle owner, as that would mean we would get duplicated states sent over to the callback from as many lifecycle owners as the callback is attached to. Instead, the [ProcessLifecycleOwner](https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner) takes care of notifying each state once (and the ON_CREATE one is only provided once throughout the whole app process regardless of how many screens are open). So for example, in the case that many lifecycle owners are on a screen at a time, e.g. an Activity and several Fragments, and the home button is pressed, then the states ON_PAUSE and ON_STOP would be received only once by the ProcessLifecycleOwner callback, essentially providing an app-wide lifecycle state update which is what we're looking for. On top of that, the [ProcessLifecycleOwner](https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner) is aware of screen config changes in order to avoid wrongly notifying states that would denote an app going into the background, when in reality it's only an app rotation for example, so it comes in very handy to provide an overall app state.

At first, we were pondering about just providing those 2 app states that you mentioned (foreground and background) but later we realized that those are too generic, which can cause some important information to get lost (for example, if a crash occurs after ON_PAUSE but before ON_STOP, said insight would be lost if we make the assumption that the app was either in the foreground/background, which wouldn't be helpful for someone taking a look at the recorded telemetry to find out the root cause) plus it also, as mentioned in the example, would make us to assume what exactly means that an app is in the foreground or in the background, which might seem simple enough, but at the end, we would still be removing information that might be valuable for the consumer of the data. So at the end, that's the reason we decided to go with platform-specific states that each developer from each platform would feel familiar with. And in the case of Android, the fact that for [ProcessLifecycleOwner](https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner) Google decided to reuse common Android lifecycle state names to represent the app's process one, made it a natural fit for our use case.

Reply:
Ah, sorry I should've mentioned the dedupe. Yes, no multiple events for multiple lifecycle owners firing the same thing triggered by the same app action. 

However, I think there's still some incongruity in this as these events aren't recognized as app states by us. `ProcessLifecycleOwner`, in fact, is just an convenient implementation for app devs to use to act upon the global firing of these events by lifecycle owners (of which the application isn't one). Perhaps background/foreground is too simplistic and ill-defined, but these events don't necessarily match up to the application state either. 

For instance, a warm start that doesn't result in the main activity being created (e.g. a WorkManager job waking up to do something) won't fire the ON_CREATE event, I don't think, so the app could have been initialized partially, but no activity has been created. If an Event is logged during that state, which should the value of this be? Also, START, RESUMED, PAUSED, and STOPPED aren't so much states as they are events in the activity lifecycle. While it's possible to have started but not resumed, that is a very small bit of time and really not a period that you'd want to log any events during (other than the actual callback being invoked)

In your example, if you want to find out a crash happened after ON_PAUSE but before ON_STOP,  it's probably better to log that as a breadcrumb trail of events on a Span that happens before a crash event. Plus you'll want the lifecycle owner name to do that kind of debugging, which isn't captured in the app state. 

Overall, while I think there are advantages to using these predefined events as pseudo proxies for app state, there's just enough nuance there that the mapping doesn't quite fit on Android.

Reply:
> (e.g. a WorkManager job waking up to do something) won't fire the ON_CREATE event

It does, and it actually helps when needed to get an insight of a background work in either case, cold and warm starts. Since it's possible take a look at what was the latest app state that was registered in case you've found an issue with your worker. The idea of providing the whole picture is so that people can decide how to handle that information in the best way for their use case.

> if you want to find out a crash happened after ON_PAUSE but before ON_STOP, it's probably better to log that as a breadcrumb trail of events on a Span that happens before a crash event.

I agree that would be a good solution, however, it doesn't necessarily clash with the idea of providing a full picture of what's the app state, which can be used as complementary information.

> Overall, while I think there are advantages to using these predefined events as pseudo proxies for app state, there's just enough nuance there that the mapping doesn't quite fit on Android.

I'm open to discussing the parts that don't fit on Android and the use cases that this approach can't solve. I personally think Google created this tool because to them it represents the lifecycle of an app process (hence the name) and, probably on purpose, it also contains familiar terms for Android devs, which is nice because it makes it intuitive enough. Though if there's a better alternative that also provides the whole picture in a straightforward way, both for implementation and read intuitiveness, I'm all up for it.

Reply:
Ah, so ON_CREATE is fired by `ProcessLifecycleInitializer` which is only used if you use the [Jetpack Startup App Startup library](https://developer.android.com/topic/libraries/app-startup). So `ProcessLifecycleOwners`is more or less the Google Android Jetpack app lifecycle, not the vanilla Android (AOSP) lifecycle.

Reply:
Android is, unfortunately, not as straightforward as the docs might have believe, especially if you go out of the "Android app download from the Google Play Store running on an Phone running default Android" zone. Like, how this stuff will work with embedded devices, Android Auto, or apps without access to Play Services is pretty undefined if you go off the beaten path. That blog post series is even necessary because Google doesn't really go down to that level of details in their docs - they expected that most folks don't care, and those that do will just read the source code, which is what we have to do. For app devs, it mostly doesn't matter, but when doing framework-y stuff, we are kind of obliged to at least try to make things work for all Android apps (which may not turn out to be reasonable...)

I'll let you know when I finish my data gathering. In the meantime, if you want to look at another Google-y way to classify the main app process state, [this page](https://developer.android.com/guide/components/activities/process-lifecycle) provides another viewpoint: `foreground`, `visible`, `service` and `cached` based on lifecycle owner transition events (we can add something for an app that hasn't fully initialized yet).

Anyway, I'll send the bat signal out and get back to you.

Reply:
I agree that Google’s current lifecycle names used in their ProcessLifecycleOwner tool are not perfect, but those names (sometimes all of them, such as in the case of Activities and Fragments, while sometimes a subset of them, which is the case for Services and the Application itself) are everywhere. So they definitely have the value of intuitiveness, both from the fact that they are everywhere, and also because they are reminiscent of a lifecycle, where the idea of a beginning and an end are clearly stated for all of its layers (for a creation, there’s a destroy, for a start, there’s a stop, for resume, pause).

Probably I would remove the paused and resumed ones since they don’t seem to help much from an application point of view given that there are the started and stopped ones that are called at about the same time, though it’s debatable. Though apart from that, it’s still not clear to me why the rest of the names would cause any issues or wouldn’t solve problems that other names such as `foreground`, `visible`, `service` and `cached` would when it comes to describing the lifecycle of an app. From a first glance at those alternatives, it doesn’t seem too clear what could be the difference between `foreground` and `visible` for example, also `service` doesn’t seem to refer to the state of an app and `cached` sounds more about what’s in disk regarding that app, rather what its lifecycle state might be, not to mention that, despite having a `foreground`, there doesn’t seem to be an opposite to it, same for `visible`.

Now, after taking a look at the docs in the link you provided, I could get a better understanding of what those names mean (interestingly enough, some of the explanations provided there refer to the common lifecycle states in order to explain what they mean, with phrases like “its onResume() method has been called”), and also what are they meant to be used for, which seems to be something that the OS checks in order to decide what processes to kill, so they do look like states, though probably not lifecycle ones and more of “can or can not be destroyed” kind of states, not to mention that since their use case seems to be oriented to the OS internal processes, it’s not something regular Android devs might be familiar with, definitely not as with the ones proposed in this PR.

So based on the above, if the names we choose to represent the states of an app have to be documented in a way that we have to refer to the more common intuitive ones, that doesn’t sound sensible, because in the end we’d be just creating extra steps for both the implementation and the consumers of this data, which also defeats the purpose of the reason why different state names are provided for Android and iOS, so if we have to come up with custom names, we might as well align both platforms and have mappings depending on each common state names. 

My ultimate concern about all this is that, from re-reading the messages in this thread, I still find it difficult to understand what’s the problem with using common lifecycle state names in Android. What I could get is that there seems to be a concern about “gaps” with this naming that might cause issues for some types of apps. Though it sounds like it’s also not sure that those issues would ever arise, so, in the end, it looks like we might be trying to solve some problems that we don’t know will ever exist, and the reason why this is concerning to me is that this starts to look like a case of “premature optimization” issue which, in my experience, tends to delay, and sometimes even fully kill any progress.

So my conclusion for this very long message is that I believe that the benefits from using a familiar set of names are very important for the overall usability of a feature, even more so when an official tool provides them in a convenient way, and some of the names actually make sense, so trying to come up with different naming, just for the sake of not reusing the ones that are already common, doesn’t sound sensible, at least not without an important reason of an existing problem, otherwise we’d be basically swimming against the current based on assumptions.

One way I think this can be addressed would be to launch this spec as `not-stable`, which I believe is a common practice in other Otel specs, and then if after some time some people start finding issues with it, we can adjust it accordingly. But trying to get ahead of every potential issue only sounds like a recipe to get paralyzed and probably over-complicate things.

Reply:
I appreciate all your thought into this. I actually wrote a spec last night explain a lot of this further, but I clicked the wrong button and lost it all. Instead of rewriting it, let me explain my main concerns with the current proposal more succinctly and present an alternate, simplified one:

- The current states are basically proxies for the state of the current Activity, but not all Android apps are driven by a current Activity (some don't even have one). 
- `Resume` and `Paused` aren't events that exist for Services, Broadcast Receivers, or even the Application. For apps that drive widgets or do their work without a UI, they will always be in the `started` state when active while for apps with Activities, they will mostly be in `resumed` when they are being used. This is odd given the description of the `started` state as being "about to be shown"

If the goal is for Android devs to have a familiar set of names and to not prematurely optimize, we should use terminology that we are familiar with, namely `foreground` (i.e. app UI is in focus) and `background` (i.e. app UI is not visible or in focus, even if it's visible). To cover the other times when we may want to log an event, we can add `launching` or `created` (i.e. the state where we are creating a new instance of the app). Effectively, this mashes together started/resumed and paused/stopped into terms that are applicable with or without activities. Note: `onDestroy` is not guaranteed to be called for Lifecycle owners, and definitely will not be called for the app process itself, so we could probably leave this out for Android.

These terms will achieve the goal of being familiar and not over complicate things (e.g. we can break these down further...)

Reply:
> The current states are basically proxies for the state of the current Activity

While Activities most likely were the most popular class where these methods had to be overridden, I believe stating that these essentially belong only to Activities is not accurate, and while some of those methods aren't fully present in places where there's some idea of a lifecycle, such as Services and the Application itself, there are still methods in those that belong to the same set (not to mention other places where all of them are present, such as Fragments), which is why I'm pretty sure if an Android dev thinks of adding logic in a place where something is being created, it'd be natural to think of overriding some onCreate method, which is present in all those classes I mentioned, and is the reason why I believe is just a natural way to visualize lifecycles in Android in general.

> but not all Android apps are driven by a current Activity (some don't even have one).

I'm probably not up-to-date with all the changes that the Android development tools have gotten recently, so maybe there's something I could be missing here, however, if by this sentence you're referring to apps that only use Fragments for their screens, those fragments always have to have a host Activity, so it's something that has to be present, even if it's not directly used (in case your concern would be that tools such as [ProcessLifecycleOwner](https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner) can only work when there are activities being used in the app).

Regarding the names you suggest, just to put them in perspective with the ones proposed here to confirm if I understood them correctly:

- `created` -> `launching` or `created`
- `started` -> `foreground`
- `resumed` -> N/A
- `paused` -> N/A
- `stopped` -> `background`
- `destroyed` -> N/A

Is that correct?

Reply:
>  I believe stating that these essentially belong only to Activities is not accurate

I think `create` is shared by all, and there's some actions of starting/stopping for all of them, but the notion of resumed and paused are really implementation details for Activities and Fragments. I think my main concern is the splitting of those two (started and resumed) and there being little difference in practice, and to a lesser extent, paused and stopped.

> if by this sentence you're referring to apps that only use Fragments for their screens

No, I'm talking about an app that relies on Services to power widgets or non-visible work. Embedded Android apps would fit in this case as well. This is what the `service` process state as supposed to denote.

And yes, your mapping is what I was trying to express (maybe `resumed` -> `foreground` instead in term of the description). This simplified set of states is still imperfect for apps with no Activities (e.g. an app with no UI is still "foreground" when it's active?) but I think it best achieves the goal of using familiar terminology and not having too many states (i.e. `active-but-invisible`)

Reply:
Just so we can compare them side by side, I've outlined the 3 proposals we've talked around with updated OS condition and descriptions to be more accurate for Android. Proposal 3 is basically like your current proposal, but with updated OS terminology and descriptions and without the `destroyed` state. Proposal 2 is like Proposal 3, but with started/resumed and paused/stopped collapsed together so valid states and state transitions don't depend on whether the app has an Activity. Proposal 1 is my preference, as it doesn't reuse terminology for the Activity lifecycle but is still descriptive and familar to Android devs.

Let me know what you think!

**Proposal 1: Familiar terminology, fewer states**

| Name         | OS terminology  | description                                      |
|--------------|-----------------|--------------------------------------------------|
| `created`    | [1]             | The app has been launched                        |
| `foreground` | [2]             | The app is in the foreground or otherwise active |
| `background` | [3]             | The app is in the background                     |

[1] Any time before Activity.onResume() or, if the app has no Activity, Context.startService() has been called in the app for the first time
[2] Any time after Activity.onResume() or, if the app has no Activity, Context.startService() has been called when the app was in either the `created` or `background` states
[3] Any time after Activity.onPause() or, if the app has no Activity, Context.stopService() has been called when the app was in the `foreground` state

**Proposal 2: Terminology matching activity/service lifecycle, fewer states**

| Name         | OS terminology  | description                                      |
|--------------|-----------------|--------------------------------------------------|
| `created`    | [1]             | The app has been launched                        |
| `started`    | [2]             | The app is in the foreground or otherwise active |
| `stopped`    | [3]             | The app is in the background                     |

[1] Any time before Activity.onResume() or, if the app has no Activity, Context.startService() has been called in the app for the first time
[2] Any time after Activity.onResume() or, if the app has no Activity, Context.startService() has been called when the app was in either the `created` or `stopped` states
[3] Any time after Activity.onPause() or, if the app has no Activity, Context.stopService() has been called when the app was in the `started` state

**Proposal 3: Terminology matching activity/service lifecycle, more states**

| Name         | OS terminology  | description                                                  |
|--------------|-----------------|--------------------------------------------------------------|
| `created`    | [1]             | The app has been launched                                    |
| `started`    | [2]             | The app is about to be in the foreground or otherwise active |
| `resumed`    | [3]             | The app is in the foreground or otherise active              |
| `paused`     | [4]             | The app is paused                                            |
| `stopped`    | [5]             | The app is in the background                                 |

[1] Any time before Activity.onStart() or, if the app has no Activity, Context.startService() has been called in the app for the first time
[2] Any time after Activity.onStart() or, if the app has no Activity, Context.startService() has been called when the app was in either the `created` or `stopped` states
[3] Any time after Activity.onResume() has been called when the app was in the `started` state
[2] Any time after Activity.onPause() has been called when the app was in either the `resumed` state
[3] Any time after Activity.onStop() or, if the app has no Activity, Context.stopService() has been called when the app was in the `paused` or `started` states

Reply:
The tables help a lot to get a better overview, thank you for that!

The second option I think is misleading because the names don't match the actual Android states 1:1 so it might cause misunderstandings.

The third option was actually our original proposal until @breedx-splk asked us to add the `destroyed` state, though I think it was more to have the complete set of states defined here, even though it's not guaranteed that we could record that state, at least not with the process lifecycle owner tool.

And regarding the first option, I'm personally fine with it to be honest, as I mentioned earlier, `foreground` and `background` were options we considered before creating this PR, though we discarded them due to being too generic (which causes that we then have to explain what do they really mean in Android terms, as you did with `foreground` -> `resumed` instead of `started`). However, from a practical point of view, and based on the fact that your description not only mentions activities but also services, those 3 states actually cover the use cases that I've come across so far, so from my side, I don't really have strong opinions against using that one. The only thing I used to consider not practical was the fact that we have to come back to Android terms to explain what the generic names mean, which seemed like just adding extra steps, however, I didn't think of non-ui apps back then, so the generic names actually help cover those other cases as well which I think it's nice.

So I'm up for going with the first approach, many thanks for taking the time to put this all together @bidetofevil

Reply:
Thanks for being patient with me @LikeTheSalad! 

I think approaching it from the point of view of `ProcessLifecycleOwner` is a very reasonable starting point. But I think it papers over some Android quirks and edge cases a little too much for my liking. The fact that it isn't Android AOSP and only an add-on via the Jetpack library tells me some folks at Android/Google feel similarly. (And trying to implement the state updates trivially with `ProcessLifecycleOwner` is probably going to lead to errors even for the generic case given the dispatch delays for onPause/onStop). Looking at what iOS has, I'm a bit jealous lol.

Yeah, the reason I prefer #1 is that it's the simplest and east to intuit, even for non-devs. The fact that Android is unspecific about app/process state in general (along with implementation messiness) means we either have to have a lot of states that may be difficult to tersely describe via name anyway, or we have to go deeper in explaining in the docs, which I think we'd have to do anyway given the Service edge case. And as you said, if the community feels differently, we can always change it to add or rename the states, which is easier than going the other way and deprecating superfluous ones.

I appreciate you listening to someone who is essentially a rando 😅 - feel free to reach out if you need anything else from me regarding this. Whatever y'all end up deciding on using, some sort of state transition diagram in the docs might be useful to even more clearly illustrate this (literally...)

Reply:
I agree, we're both jealous of iOS in this regard 😄 though overall I'm pretty happy with the outcome of this thread so many thanks for your patience as well and for your time! I hope we can continue working together on future OTel Android-related issues, which I believe will be a lot because it seems like the Android-related work in OTel will start to grow along with the recent Splunk SDK donation so, looking forward to it 🙌

Reply:
For sure. I've joined the Cloud Native Computing Foundation Slack and am posting on #otel-client-side-telemetry channel. Looking forward to working on this stuff with y'all!